File: src/main/java/org/athenian/helloworld/HelloWorldServer.java
package org.athenian.helloworld;

import io.grpc.Server;
import io.grpc.ServerBuilder;

import java.io.IOException;

public class HelloWorldServer {
    private Server server;

    public static void main(String[] args)
        throws IOException, InterruptedException {
        final HelloWorldServer server = new HelloWorldServer();
        server.start();
        server.blockUntilShutdown();
    }

    private void start()
        throws IOException {
        /* The port on which the server should run */
        int port = 50051;
        server = ServerBuilder.forPort(port)
            .addService(new GreeterImpl())
            .build()
            .start();
        System.out.printf("Server started, listening on %d%n", port);
        Runtime.getRuntime().addShutdownHook(
            new Thread(() -> {
                // Use stderr here since the logger may have been reset by its JVM shutdown hook.
                System.err.println("*** shutting down gRPC server since JVM is shutting down");
                HelloWorldServer.this.stop();
                System.err.println("*** server shut down");
            }));
    }

    private void stop() {
        if (server != null) {
            server.shutdown();
        }
    }

    /**
     * Await termination on the main thread since the grpc library uses daemon threads.
     */
    private void blockUntilShutdown()
        throws InterruptedException {
        if (server != null) {
            server.awaitTermination();
        }
    }
}

File: src/main/kotlin/org/athenian/kotlin_helloworld/withoutCR/HelloWorldServer.kt
package org.athenian.kotlin_helloworld.withoutCR

import io.grpc.ServerBuilder

class HelloWorldServer(val port: Int) {
    private val server = ServerBuilder.forPort(port).addService(GreeterImpl()).build()

    private fun start() {
        server.start()
        println("Server started, listening on $port")
        Runtime.getRuntime().addShutdownHook(
            Thread {
                // Use stderr here since the logger may have been reset by its JVM shutdown hook.
                System.err.println("*** shutting down gRPC server since JVM is shutting down")
                server.shutdown()
                System.err.println("*** server shut down")
            })
    }

    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            val port = System.getenv("PORT")?.toInt() ?: 50051
            with(HelloWorldServer(port)) {
                start()
                server.awaitTermination()
            }
        }
    }
}
